src/contract.rs

```diff
@@ -491,9 +541,26 @@ fn try_redeem<S: Storage, A: Api, Q: Querier>(
     env: Env,
     amount: Uint128,
 ) -> StdResult<HandleResponse> {
-    let sender_address = deps.api.canonical_address(&env.message.sender)?;
     let amount_raw = amount.u128();
 
+    let mut config = Config::from_storage(&mut deps.storage);
+    let constants = config.constants()?;
+    if !constants.redeem_is_enabled {
+        return Err(StdError::generic_err(
+            "Redeem functionality is not enabled for this token.",
+        ));
+    }
+    let total_supply = config.total_supply();
+    if let Some(total_supply) = total_supply.checked_sub(amount_raw) {
+        config.set_total_supply(total_supply);
+    } else {
+        return Err(StdError::generic_err(
+            "You are tyring to redeem more tokens than what is available in the total supply",
+        ));
+    }
```

Comment:
putting this before trying to subtract from the user's balance allows users to circumvent a private total supply by trying to redeem more than they have. within 10-20 txs you can get a very good idea of what the exact number is.

Reply:
Ah ok.  I moved it up to save in gas cost.  If the total supply check is after the balance check you would have to reload the Config, because it is a storage wrapper, and you grab a &mut to storage when doing the balance processing.  So, we could move the redeem config check to after the balance processing so that we only have to load the Config once, and just accept that someone trying a redeem when it is disabled will spend a little more gas than necessary because it actually changes his balance before it checks that redeem is disabled.  Since the error because it is disabled will revert the balance change, there's no harm.  I'd probably rather have the person making the error pay more gas than everyone dong a redeem pay more gas because of an extraneous Config load.

Reply:
So, the price of `Config::from_storage` and `ReadonlyConfig::from_storage` is negligible as it doesn't actually access the storage (unlike say, `AppendStore`).
I will say though that i regret the design decision i made of holding the storage object inside other types. Given the cosmwasm_std approach of only providing a reference to the storage, a design with a lot less friction is just requiring the reference to storage as an argument to methods on the type.
(Long term i hope we'll get a new cosmwasm_std/secret_std which provides queries with an owned `Store` type, and handles/init with an owned `StoreMut` type, both of whom can easily be `Copy` or at the very least `Clone`)